lib/active_merchant/billing/gateways/opp.rb

```diff
@@ -0,0 +1,399 @@
+    #   Because this list is much bigger than Gateway::STANDARD_ERROR_CODE, only fraction is mapped to Gateway::STANDARD_ERROR_CODE. 
+    #   All other codes are mapped as Gateway::STANDARD_ERROR_CODE[:processing_error], so if this is the case, 
+    #   you may check the original result code from OPP that can be found in response.params['result']['code']
+    #       
+    # == Special features
+    #   For purchase method risk check can be forced when options[:riskWorkflow] = true 
+    #   This will split (on OPP server side) the transaction into two separate transactions: authorize and capture, 
+    #   but capture will be executed only if risk checks are successful.   
+    #
+    #   For testing you may use the test account details listed fixtures.yml under opp. It is important to note that there are two test modes available:
+    #     options[:testMode]='EXTERNAL' causes test transactions to be forwarded to the processor's test system for 'end-to-end' testing
+    #     options[:testMode]='INTERNAL' causes transactions to be sent to opp simulators, which is useful when switching to the live endpoint for connectivity testing.
+    #   If no testMode parameter is sent, testMode=INTERNAL is the default behaviour.
+    #
+    #   Shopping Cart, Billing Address, Shipping Address, Custom Parameters are supported as described under https://docs.oppwa.com/parameters
+    #   See complete example above for details. 
+    #
+    #   == Tokenization
+    #  When createRegistration is set to true, the payment details will be stored and a token will be returned in registrationId response field, 
+    #  which can subsequently be used to referenced the stored payment.
+
+      self.test_url = 'https://test.oppwa.com/v1/payments'
+      self.live_url = 'https://oppwa.com/v1/payments'
+
+      self.supported_countries = %w(AD AI AG AR AU AT BS BB BE BZ BM BR BN BG CA HR CY CZ DK DM EE FI FR DE GR GD GY HK HU IS IN IL IT JP LV LI LT LU MY MT MX MC MS NL PA PL PT KN LC MF VC SM SG SK SI ZA ES SR SE CH TR GB US UY)
+      self.default_currency = 'EUR'
+      self.supported_cardtypes = [:visa, :master, :american_express, :diners_club, :discover, :jcb, :maestro, :dankort] 
+
+      self.homepage_url = 'https://docs.oppwa.com'
+      self.display_name = 'Open Payment Platform'
+
+      METHOD_TO_PAYMENTTYPE_MAPPING = { purchase: 'DB', authorize: 'PA', capture: 'CP', refund: 'RF', void: 'RV', pacp: 'PA.CP' }
+
+      def initialize(options={})
+        requires!(options, :userId, :password, :entityId)
+        super
+      end
+
+      def purchase(money, payment, options={})
+        # debit
+        execute_dbpa(options[:riskWorkflow] ? METHOD_TO_PAYMENTTYPE_MAPPING[:pacp]: METHOD_TO_PAYMENTTYPE_MAPPING[__method__], 
+          money, payment, options)
+      end
+
+      def authorize(money, payment, options={})
+        # preauthorization PA
+        execute_dbpa(METHOD_TO_PAYMENTTYPE_MAPPING[__method__], money, payment, options)
```

Comment:
Not a fan of using `__method__` like this. Just replace with an explicit reference to the method name.